sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/samplers/SamplingResult.scala

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk.trace.samplers
+
+import cats.Hash
+import cats.Show
+import cats.syntax.show._
+import org.typelevel.otel4s.sdk.Attributes
+import org.typelevel.otel4s.trace.TraceState
+
+/** Sampling result returned by [[Sampler.shouldSample]].
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/trace/sdk/#shouldsample]]
+  */
+sealed trait SamplingResult {
+
+  /** The decision on whether a span should be recorded, recorded and sampled or
+    * not recorded.
+    */
+  def decision: SamplingDecision
+
+  /** The attributes that will be attached to the span.
+    */
+  def attributes: Attributes
+
+  /** A modifier of the parent's TraceState.
+    *
+    * It may return the same trace state that was provided originally, or an
+    * updated one.
+    *
+    * '''Note''': If an empty trace state is returned, the trace state will be
+    * cleared, so samplers should normally return the passed-in trace state if
+    * they do not intend to change it.
+    */
+  def traceStateUpdater: SamplingResult.TraceStateUpdater
```

Comment:
Java SDK uses a default method in the interface: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/SamplingResult.java#L134

Unfortunately, this approach does not work if we want to keep the trait sealed.

I've added `TraceStateUpdater` instead.

Reply:
> Unfortunately, this approach does not work if we want to keep the trait sealed.

Dumb question: why? I'm having trouble seeing the difference.

Reply:
According to the spec, the `SamplingResult` could be defined as:
```scala
sealed trait SamplingResult {
  def decision: SamplingDecision
  def attributes: Attributes
  def updatedTraceState: TraceState
}
```

But with this design, we cannot use constant sampling results (e.g. `SamplingResult.Drop`) anymore, because we always need to return the parent's trace state. So, we need to create a new sampling result on every call of `Sampler.shouldSample`.

Since only a few samplers in the **contrib** module (e.g. [ConsistentSampler](https://github.com/open-telemetry/opentelemetry-java-contrib/blob/4a00e328fe8d40c31b2552d65faf54ff48e42c67/consistent-sampling/src/main/java/io/opentelemetry/contrib/sampler/consistent/ConsistentSampler.java#L294)) provide a custom implementation of the `getUpdatedTraceState` Java SDK team decided to go their way, I assume.


____

We can still use a method, but encoding will be cumbersome:


```scala
sealed trait SamplingResult {
  ...
  def updateTraceState(state: TraceState): TraceState
}

def apply(
    decision: SamplingDecision,
    attributes: Attributes,
    updateState: TraceState => TraceState
): SamplingResult = ...
```

When we use a function instead of a `TraceStateUpdater` ADT, it's more complex to:
1) Implement a proper hashing
2) Informative toString
3) Could be less intuitive